logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java

```diff
@@ -0,0 +1,196 @@
+import io.camunda.zeebe.scheduler.ActorCondition;
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The sequencer is a multiple-producer, single-consumer queue of {@link LogAppendEntry}. It buffers
+ * a fixed amount of entries and rejects writes when the queue is full. The consumer may read at its
+ * own pace by repeatedly calling {@link Sequencer#tryRead()} or register for notifications when new
+ * entries are written by calling {@link Sequencer#registerConsumer(ActorCondition)}.
+ *
+ * <p>The sequencer assigns all entries a position and makes that position available to its
+ * consumer. The sequencer does not copy or serialize entries, it only keeps a reference to them
+ * until they are handed off to the consumer.
+ */
+public final class Sequencer implements LogStreamBatchWriter, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(Sequencer.class);
+  private final int partitionId;
+  private final int maxFragmentSize;
+
+  private volatile long position;
+  private volatile boolean isClosed = false;
+  private volatile ActorCondition consumer;
+  private final Queue<SequencedBatch> queue = new ArrayBlockingQueue<>(128);
+  private final ReentrantLock lock = new ReentrantLock();
+  private final SequencerMetrics metrics;
+
+  public Sequencer(final int partitionId, final long initialPosition, final int maxFragmentSize) {
+    LOG.trace("Starting new sequencer at position {}", initialPosition);
+    this.position = initialPosition;
+    this.partitionId = partitionId;
+    this.maxFragmentSize = maxFragmentSize;
+    this.metrics = new SequencerMetrics(partitionId);
+  }
+
+  /**
+   * @param eventCount the potential event count we want to check
+   * @param batchSize the potential batch Size (in bytes) we want to check
+   * @return True if the serialized batch would fit within {@link Sequencer#maxFragmentSize}, false
+   *     otherwise.
+   */
+  @Override
+  public boolean canWriteEvents(final int eventCount, final int batchSize) {
+    final int framedMessageLength =
+        batchSize
+            + eventCount * (DataFrameDescriptor.HEADER_LENGTH + FRAME_ALIGNMENT)
+            + FRAME_ALIGNMENT;
+    return framedMessageLength <= maxFragmentSize;
+  }
+
+  /**
+   * @param appendEntry the entry to write
+   * @param sourcePosition a back-pointer to the record whose processing created this entry
+   * @return -1 if write was rejected, the position of the entry if write was successful.
+   */
+  @Override
+  public long tryWrite(final LogAppendEntry appendEntry, final long sourcePosition) {
+    if (isClosed) {
+      LOG.warn("Rejecting write of {}, sequencer is closed", appendEntry);
+      return -1;
+    }
+    final boolean isEnqueued;
+    final long currentPosition;
+
+    lock.lock();
```

Comment:
> So I would propose enabling fairness
Even though `ReentrantLock` points out lower throughput and limited effectiveness because thread scheduling is not fair? I think it might be worth a try but I'm not sure we could even tell if this has any effect or even if we have starved writer threads at all.

> and also locking interruptibly
I'm not sure I understand what we would get from that. Is anything in Zeebe ever interrupting writer threads? Or is it just a matter of "in theory this is better"?

Reply:
It's mostly a defensive technique, when doing asynchronous programming you want to avoid possible blocking points. In the case of a lock, it's mostly about deadlocks (though high latency variance between task execution isn't great either), which hopefully you never run into, but this is about preventing this issue (e.g. someone changes some other usage and we end up with this being blocking).

Anyway, as we discussed, I'd rather invest time in removing the lock entirely :+1: